### ISSUE TITLE
UWSM Fails to Start Compositor For the First Time After Booting 

### ISSUE THREAD

**Issue by ShadowS0ng** [Vladimir-csp/uwsm#76]
when i try to start compositor with uwsm it logs in for an instant and immediately logs out. and i have to start it once again. the problem doesn't exist if i use a **Graphical Display-Manager** like sddm but it's there when using something like greetd/agreety or without display-manager it also is non-existent on a VM i created for testing this same problem. 

So far i set `UWSM_WAIT_VARNAME_SETTLETIME` to 10 and added `graphical.target` to `getty@.service` wants without success .

Is there any fix to this? 

below is a relevant part of journalctl if it helps:
```
Dec 20 19:35:26 archlinux uwsm_waitenv[1943]: Expecting variables to appear in activation environment:
Dec 20 19:35:26 archlinux uwsm_waitenv[1943]:   HYPRLAND_INSTANCE_SIGNATURE, WAYLAND_DISPLAY
Dec 20 19:35:26 archlinux uwsm_hyprland.desktop[1944]: Starting: Hyprland -c /home/AX/.config/hypr/hyprlandUwsm.conf...
Dec 20 19:35:26 archlinux uwsm_hyprland.desktop[1946]: Expecting variables to appear in activation environment:
Dec 20 19:35:26 archlinux uwsm_hyprland.desktop[1946]:   WAYLAND_DISPLAY, HYPRLAND_INSTANCE_SIGNATURE
Dec 20 19:35:26 archlinux uwsm_start[1897]: Holding until PID 1944 exits
Dec 20 19:35:34 archlinux uwsm_waitenv[1943]: Still waiting for variables in activation environment, nearing timeout:
Dec 20 19:35:34 archlinux uwsm_waitenv[1943]:   HYPRLAND_INSTANCE_SIGNATURE, WAYLAND_DISPLAY
Dec 20 19:35:34 archlinux uwsm_hyprland.desktop[1946]: Still waiting for variables in activation environment, nearing timeout:
Dec 20 19:35:34 archlinux uwsm_hyprland.desktop[1946]:   WAYLAND_DISPLAY, HYPRLAND_INSTANCE_SIGNATURE
Dec 20 19:35:36 archlinux systemd[670]: wayland-wm@hyprland.desktop.service: start operation timed out. Terminating.
Dec 20 19:35:36 archlinux systemd[670]: wayland-session-waitenv.service: start operation timed out. Terminating.
Dec 20 19:35:36 archlinux systemd[670]: wayland-session-waitenv.service: Main process exited, code=killed, status=15/TERM
░░ Subject: Unit process exited
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░
░░ An ExecStart= process belonging to unit UNIT has exited.
░░
░░ The process' exit code is 'killed' and its exit status is 15.
Dec 20 19:35:36 archlinux systemd[670]: wayland-session-waitenv.service: Failed with result 'timeout'.
░░ Subject: Unit failed
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░
░░ The unit UNIT has entered the 'failed' state with result 'timeout'.
Dec 20 19:35:36 archlinux systemd[670]: Failed to start Wait for WAYLAND_DISPLAY and other variables.
░░ Subject: A start job for unit UNIT has failed
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░
░░ A start job for unit UNIT has finished with a failure.
░░
░░ The job identifier is 86 and the job result is failed.
Dec 20 19:35:36 archlinux systemd[670]: wayland-session-waitenv.service: Triggering OnFailure= dependencies.
Dec 20 19:35:36 archlinux systemctl[681]: Job for wayland-wm@hyprland.desktop.service canceled.
Dec 20 19:35:36 archlinux login[661]: pam_unix(login:session): session closed for user AX.
```

**Comment by Vladimir-csp** [comment]
Strange, Hyprland usually puts variables to activation environment itself, unless `HYPRLAND_NO_SD_VARS=1` is set. What versions of Hyprland and UWSM do you use? There was a time when this inhibition was used, but judging by `uwsm_waitenv` presense, it's not that version.
Anyway, if you add `exec-once = exec uwsm finalize` to Hyprland config, it should work.

**Comment by ShadowS0ng** [comment]
> Strange, Hyprland usually puts variables to activation environment itself, unless `HYPRLAND_NO_SD_VARS=1` is set. What versions of Hyprland and UWSM do you use? There was a time when this inhibition was used, but judging by `uwsm_waitenv` presense, it's not that version. Anyway, if you add `exec-once = exec uwsm finalize` to Hyprland config, it should work.

My bad, forgot to mention i also added `exec-once = exec uwsm finalize` without success also added `WAYLAND_DISPLAY` and `HYPRLAND_INSTANCE_SIGNATURE` as arguments and it didn't work. I'm using latest git version. the one on VM was latest release version in Arch repos though. my laptop is an i5 Ivy-bridge so i suspect it's slow and some things doesn't happen in a timely manner.

**Comment by Vladimir-csp** [comment]
`uwsm finalize` should produce this message in the journal: `uwsm_hyprland.desktop[3175]: Exporting variables to systemd user manager:`
In general, any stdout from `exec-once` or `exec` dispatchers should be visible under `uwsm_hyprland.desktop` tag. Does anything run?

**Comment by ShadowS0ng** [comment]
> `uwsm finalize` should produce this message in the journal: `uwsm_hyprland.desktop[3175]: Exporting variables to systemd user manager:` In general, any stdout from `exec-once` or `exec` dispatchers should be visible under `uwsm_hyprland.desktop` tag. Does anything run?

> `uwsm finalize` should produce this message in the journal: `uwsm_hyprland.desktop[3175]: Exporting variables to systemd user manager:` In general, any stdout from `exec-once` or `exec` dispatchers should be visible under `uwsm_hyprland.desktop` tag. Does anything run?

`Exporting variables to systemd user manager:` happens twice : 

```
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[509990]: All expected variables appeared in activation environment.
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]: Marking variables for later cleanup from systemd user manager on stop:
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   HYPRCURSOR_SIZE
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   HYPRCURSOR_THEME
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   HYPRLAND_CMD
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   HYPRLAND_INSTANCE_SIGNATURE
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]: Exporting variables to systemd user manager:
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   DISPLAY (already set)
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   HYPRCURSOR_SIZE
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   HYPRCURSOR_THEME
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   HYPRLAND_CMD
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   HYPRLAND_INSTANCE_SIGNATURE (already set)
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   WAYLAND_DISPLAY (already set)
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   XCURSOR_SIZE
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]:   XCURSOR_THEME
Dec 21 15:14:33 archlinux uwsm_hyprland.desktop[510023]: Declaring unit for hyprland.desktop ready.
Dec 21 15:14:33 archlinux systemd[508856]: Started Main service for Hyprland, An intelligent dynamic tiling Wayland compositor.
░░ Subject: A start job for unit UNIT has finished successfully
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░
░░ A start job for unit UNIT has finished successfully.
░░
░░ The job identifier is 59.
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]: Exporting variables to systemd user manager:
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]:   DISPLAY (already set)
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]:   HYPRCURSOR_SIZE (already set)
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]:   HYPRCURSOR_THEME (already set)
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]:   HYPRLAND_CMD (already set)
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]:   HYPRLAND_INSTANCE_SIGNATURE (already set)
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]:   QT_QPA_PLATFORMTHEME (already set)
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]:   WAYLAND_DISPLAY (already set)
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]:   XCURSOR_SIZE (already set)
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]:   XCURSOR_THEME (already set)
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]: Marking variables for later cleanup from systemd user manager on stop:
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]:   QT_QPA_PLATFORMTHEME
Dec 21 15:14:34 archlinux uwsm_hyprland.desktop[509990]: Unit for hyprland.desktop is already active.
```
Non of them have `[3175]` even though i have the `exec-once = exec uwsm finalize` line in my config.
This is a successful start BTW meaning it's not first time after booting. In the unsuccessful runs it doesn't happen

**Comment by Vladimir-csp** [comment]
`[3175]` is a PID, so it is different every time.
This output looks normal. One set of messages comes from `uwsm finalize`, the other comes from forked env watcher that ensures variables are properly marked and duplicated for dbus environment. Hyprland correctly exports its vars just before execs happen, that's why `(already set)` mark appears.

Please post full log of unsuccessful attempt, from login to unit timeout.

**Comment by ShadowS0ng** [comment]
> `[3175]` is a PID, so it is different every time. This output looks normal. One set of messages comes from `uwsm finalize`, the other comes from forked env watcher that ensures variables are properly marked and duplicated for dbus environment. Hyprland correctly exports its vars just before execs happen, that's why `(already set)` mark appears.
> 
> Please post full log of unsuccessful attempt, from login to unit timeout.

> `[3175]` is a PID, so it is different every time. 
Ah. I see . 

[This](https://github.com/user-attachments/files/18218675/journalOutput1.log) is the log file. The first 1053 lines are kernel lines.

**Comment by ShadowS0ng** [comment]
[This](https://github.com/user-attachments/files/18219191/journalOutput2.log) is also another one which is from log in to tty to failed attempt

**Comment by Vladimir-csp** [comment]
Looks like nothing is happening for 10 seconds after compositor launch.
If your system is that slow, you can try to adjust start timeout: `systemctl --user edit wayland-wm@.service`
```
[Service]
TimeoutStartSec=30
```

**Comment by ShadowS0ng** [comment]
> Looks like nothing is happening for 10 seconds after compositor launch. If your system is that slow, you can try to adjust start timeout: `systemctl --user edit wayland-wm@.service`
> 
> ```
> [Service]
> TimeoutStartSec=30
> ```

I tried that without success. However I Installed hyprland and uwsm on an Arch Live iso and it started without any problem so i suppose there's some thing wrong with my current installation maybe some driver problems, some systemd services i have enabled or whatever else, and i want to thank you both for this project and the time you put to help with this problem. also It's not only with hyprland the sway is the same. Is there any chance it's related to the use of btrfs?

**Comment by ShadowS0ng** [comment]
## Update
I figured out it has something to do with font cache probably. 
I added `fc-cache -fv > /dev/null 2>&1` to the script that runs uwsm-hyprland and it works fine now.

**Comment by Vladimir-csp** [comment]
Do you have any idea why font cache can affect startup? Is it a delay issue (`fc-cache` just taking a lot of time and resources preventing anything else to start in time)?

**Comment by ShadowS0ng** [comment]
> Do you have any idea why font cache can affect startup? Is it a delay issue (`fc-cache` just taking a lot of time and resources preventing anything else to start in time)?

I think it's not the time delay that it adds because when i use `fc-cache >/dev/null 2>&1` which doesn't have the -f flag it also works.
One other thing i realized is when i set `nvidia_drm.modeset=1 fbdev=1` kernel parameters (I have two GPUs) and install `mesa-amber` package instead of `mesa`, I can start it without `fc-cache` with no problem.Although it won't work if i set the modesetting through /etc/modeprobe.d/.  I don't get what's the exact reason though.

**Comment by ArtemSmaznov** [comment]
I have also been having the same issue ever since I switched to using UWSM. While getting used to this flow I've settled into logging to TTY and launching via script:
```
# ~/.local/bin/h
uwsm start hyprland.desktop
```
as with the others, first launch it segfaults, second one immediately after works fine. One thing I've noticed though that I haven't seen mentioned before is that if I wait for a bit after logging in before launching the script, then it works fine on 1st invocation. Haven't timed it but the wait should be somewhere between 1-5min.

I've updated my script to the following now and it resolves the issue for me as well:
```
# ~/.local/bin/h
fc-cache > /dev/null 2>&1
uwsm start hyprland.desktop
``` 

OS: Arch Linux x86_64
Kernel: Linux 6.13.5-zen1-1-zen
GPU: AMD Radeon RX 7800 XT [Discrete]
Hyprland 0.47.2
UWSM 0.21.1-1
Syslog [uwsm-1st-launch-crash.log](https://github.com/user-attachments/files/19039653/uwsm-1st-launch-crash.log)

Let me know if you need more info or if I should open a separate issue for my case.

**Comment by Random-Penguin** [comment]
I am also getting this issue; however, I have tried all of the solutions on this page without success.
Here's the relevant part of the journal:
```
Jul 21 12:26:22 randomarch systemd[558]: Starting Wait for WAYLAND_DISPLAY and other variables...
Jul 21 12:26:22 randomarch systemd[558]: Starting Main service for Hyprland, An intelligent dynamic tiling Wayland compositor...
Jul 21 12:26:22 randomarch uwsm_hyprland.desktop[713]: Starting: Hyprland...
Jul 21 12:26:22 randomarch uwsm_waitenv[712]: Expecting variables to appear in activation environment:
Jul 21 12:26:22 randomarch uwsm_waitenv[712]:   WAYLAND_DISPLAY, HYPRLAND_INSTANCE_SIGNATURE
Jul 21 12:26:22 randomarch uwsm_hyprland.desktop[715]: Expecting variables to appear in activation environment:
Jul 21 12:26:22 randomarch uwsm_hyprland.desktop[715]:   HYPRLAND_INSTANCE_SIGNATURE, WAYLAND_DISPLAY
Jul 21 12:26:22 randomarch uwsm[687]: waitpid: Holding login session 5 open until PID 713 exits
Jul 21 12:26:29 randomarch uwsm_waitenv[712]: Still waiting for variables in activation environment, nearing timeout:
Jul 21 12:26:29 randomarch uwsm_waitenv[712]:   WAYLAND_DISPLAY, HYPRLAND_INSTANCE_SIGNATURE
Jul 21 12:26:29 randomarch uwsm_hyprland.desktop[715]: Still waiting for variables in activation environment, nearing timeout:
Jul 21 12:26:29 randomarch uwsm_hyprland.desktop[715]:   HYPRLAND_INSTANCE_SIGNATURE, WAYLAND_DISPLAY
Jul 21 12:26:32 randomarch systemd[558]: wayland-session-waitenv.service: start operation timed out. Terminating.
Jul 21 12:26:32 randomarch systemd[558]: wayland-wm@hyprland.desktop.service: start operation timed out. Terminating.
Jul 21 12:26:32 randomarch systemd[558]: wayland-session-waitenv.service: Main process exited, code=killed, status=15/TERM
Jul 21 12:26:32 randomarch systemd[558]: wayland-session-waitenv.service: Failed with result 'timeout'.
Jul 21 12:26:32 randomarch systemd[558]: Failed to start Wait for WAYLAND_DISPLAY and other variables.
```

Also here's my `fastfetch` output:
```
  █▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀█  Hardware
  █                                                █  ├─── Machine: 81Y6 Lenovo Legion 5 15IMH05H
  █                  .,-:;//;:=,                   █  ├──────── OS: Arch Linux on x86_64
  █              . :H@@@MM@M#H/.,+%;,              █  ├──── Kernel: Linux 6.15.9-arch1-1
  █           ,/X+ +M@@M@MM%=,-%HMMM@X/,           █  ├────── Disk: ▸ ██████████ ◂ Arch-root (ext4), 33.43 GiB/453.46 GiB
  █         -+@MM; $M@@MH+-,;XMMMM@MMMM@+-         █  ╰──── Memory: ▸ ██████████ ◂ 1.10 GiB/7.66 GiB
  █        ;@M@@M- XM@X;. -+XXXXXHHH@M@M#@/.       █
  █      ,%MM@@MH ,@%=             .---=-=:=,.     █  Software
  █      =@#@@@MX.,                -%HX$$%%%:;     █  ├──────── WM: Hyprland
  █     =-./@M@M$                   .;@MMMM@MM:    █  ├───── Shell: bash 5.3.3
  █     X@/ -$MM/                    . +MM@@@M$    █  ╰── Packages: 1103
  █    ,@M@H: :@:                    . =X#@@@@-    █
  █    ,@@@MMX, .                    /H- ;@M@M=    █  Connection
  █    .H@@@@M@+,                    %MM+. %#$.    █  ├───── Wi-Fi: wlp0s20f3, up
  █     /MMMM@MMH/.                  XM@MH; =;     █  ├─── Network: Livebox-8D60 (2.4GHz, WPA2)
  █      /%+%$XHH@$=              , .H@@@@MX       █  ├── Local IP: ---------------
  █       =----------.          -%H.,@@@@@MX,      █  ╰─ Public IP: 86.228.20.203
  █       .%MM@@@HHHXX$$$%+- .:$MMX =M@@MM%.       █
  █         =XMMM@MM@MM#H;,-+HMM@M+ /MMMX=         █  Theming
  █           =%@M@M#@$-.=$@MM@@@M; %M%=           █  ├───── Theme: Tokyonight-Dark [GTK2/3/4]
  █             ,:+$+-,/H#MMMMMMM@= =,             █  ├───── Icons: Yaru-blue-dark [GTK2/3/4]
  █                   =++%%%%+/:-.                 █  ├──── Cursor: Bibata-Modern-Classic (24)
  █                                                █  ├────── Font: NotoSans Nerd Font (10pt) [GTK2/3/4]
  █▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄█  ╰─ Term Font: EnvyCodeRNFM (11pt)

                                                       Executed at: 10:23:28 on 07-08-2025 (+0200)
```

**Comment by Vladimir-csp** [comment]
What is happening before the timeout? Is something visible on screen? Do any `exec` and `exec-once` work?

**Comment by Vladimir-csp** [comment]
> It executes after the timeout with no problems

It can not execute after timeout, because after timeout compositor is terminated. Can you add `exec-once = echo testing exec-once | systemd-cat` and see if `testing exec-once` message appears in the journal.

**Comment by Vladimir-csp** [comment]
This is strange, Hyprland's log suggests that stuff is being executed, including `uwsm finalize`.

> then it doesn't see anything, it just echoes random_debug: debug into the journal.

Add `2>&1` redirection to systemctl invocation. If its stdout is empty there should be some error on stderr.

**Comment by Vladimir-csp** [comment]
Do you run this script form a neighboring login session? Maybe you are missing libpam-systemd/pam_systemd or something like that. I see a lot of tutorials and instructions still mentioning maual dbus startup via `dbus-run-session` or `dbus-launch`, even if it is not needed today as everything including login sessions is managed by systemd.

Try running this script from `exec-once` in Hyprland, units can not be missing dbus and runtime vars.

**Comment by Random-Penguin** [comment]
Oh, I was running it as a systemd service, I'll try `exec-once` soon.

**Comment by Random-Penguin** [comment]
`/etc/systemd/system/random_debug.service`

**Comment by Vladimir-csp** [comment]
It's a system unit, it will not get any user session context. Move to `~/.config/systemd/user/random_debug.service` and do `systemctl --user daemon-reload; systemctl --user add-wants wayland-wm@${your_compositor_id}.service random_debug.service`

**Comment by ArtemSmaznov** [comment]
Apologies for jumping in but this discussion sounds like this might not be a wide-spread issue but rather some edge cases. I'd like to weight in with my setup as well.

I am trying to limit the number of moving parts so launching it as simple as possible. I am logging in to TTY on boot (no display manager) and execute `exec uwsm start hyprland-uwsm.desktop` or `uwsm start hyprland-uwsm.desktop` manually. The main difference that I see is that with `exec` it logs me out after hyprland crashes while with no `exec` it just returns be back to TTY.

I have added `uwsm finalize` to hyprland startup for troubleshooting (normally I don't have it):
```conf
exec-once = exec uwsm finalize
# exec-once = exec uwsm finalize WAYLAND_DISPLAY DISPLAY
exec-once = paplay "$HOME/public/audio/windows95-startup.wav"
exec-once = uwsm app -- $HOME/.local/bin/set-wallpaper.sh
exec-once = uwsm app -S both -- eww open-many main-bar second-bar $eww-indicators $eww-playground
```

As before, first launch after boot crashes. Adding `WAYLAND_DISPLAY DISPLAY` to `finalize` doesn't seem to make any difference.

In any scenario I see `Failed to start Wait for WAYLAND_DISPLAY and other variables.` in the syslog.

And I do usually get to hear the startup sound that I have executed here via `paplay` although it doesn't play out in full as hyprland crashes midway though playing it.

And here are my syslogs for various attempts to start hyprland on first boot.

[uwsm.via-exec.with-finalize.log](https://pastebin.com/bLe1tug9)
[uwsm.via-exec.with-finalize.with-displays.log](https://pastebin.com/kbMyrjA7)
[uwsm.no-exec.with-finalize.with-displays.log](https://pastebin.com/rzttSZDU)

I am setting this up wrong? Not sure if anything here can cause this issue but just in case:
```shell
cat ~/.config/uwsm/env
# toolkit backend variables
export GDK_BACKEND=wayland,x11 # GTK: Use wayland if available, fall back to x11 if not.
export SDL_VIDEODRIVER=wayland # Run SDL2 applications on Wayland. Remove or set to x11 if games that provide older versions of SDL cause compatibility issues
export CLUTTER_BACKEND=wayland # Clutter package already has Wayland enabled, this variable will force Clutter applications to try and use the Wayland backend

# Qt variables
export QT_AUTO_SCREEN_SCALE_FACTOR=1         # (From the Qt documentation) enables automatic scaling, based on the monitor’s pixel density
export QT_QPA_PLATFORM=wayland               # QT: Use wayland if available
export QT_WAYLAND_DISABLE_WINDOWDECORATION=1 # Disables window decorations on Qt applications
export QT_QPA_PLATFORMTHEME=qt5ct            # Tells QT based applications to pick your theme from qt5ct, use with Kvantum.

# theming related variables
# export GTK_THEME=                   # Set a GTK theme manually, for those who want to avoid appearance tools such as lxappearance or nwg-look
export XCURSOR_THEME=breeze_cursors # Set your cursor theme. The theme needs to be installed and readable by your user.
export XCURSOR_SIZE=32              # Set cursor size. See here for why you might want this variable set.
```

```shell
cat ~/.config/uwsm/env-hyprland
# Set your cursor theme. The theme needs to be installed and readable by your user.
export HYPRCURSOR_THEME=HyprBibataModernClassicSVG
export HYPRCURSOR_SIZE=32 # Set cursor size. See here for why you might want this variable set.
```

**Comment by ShadowS0ng** [comment]
> > Well now it doesn't even start. systemctl --user enable random_service and then rebooting should make it start at boot, right?
> 
> See `add-wants` command example in my previous comment, it will make it start with compositor unit (stop is manual though).
> 
> > I am trying to limit the number of moving parts so launching it as simple as possible. I am logging in to TTY on boot (no display manager) and execute exec uwsm start hyprland-uwsm.desktop or uwsm start hyprland-uwsm.desktop manually. The main difference that I see is that with exec it logs me out after hyprland crashes while with no exec it just returns be back to TTY.
> 
> No need for `hyprland-uwsm.desktop` here, just plain `hyprland.desktop` will do, since you already start it via uwsm. Yes, `exec` causes the command to replace your login shell, so there is no login shell to return to. This is part of the design, otherwise a crashy compositor would defeat the purpose of a screenlock.
> 
> Without any contrasting info it all seems like something slows down the system so much that compositor can not complete startup in 10 seconds. It is not a crash, it's just a timeout. ...which can not currently be tweaked in one go if static units are used. I'll try to restore this functionality later today.
> 
> Meanwhile you can try either building uwsm with `static-units=disabled`, or, less intrusively, override `wayland-wm@.service` and `wayland-session-waitenv.service` with `TimeoutStartSec=30`, then export `UWSM_WAIT_VARNAMES_TIMEOUT=30` before starting uwsm. 30 or any other value you see fit. See how long it will really take the compositor to start. Maybe try to dump `top -b` in parallel to see what process makes the system busy. Maybe there really is some fontconfig routine that eats up all the CPU time on first login.

This actually solved the problem. 
I suspect the problem is nothing special but it's the fact that the old hardware is not fast enough to start Hyprland and daemons in 10 seconds hence it fails. This problem didn't exist when I start from a graphical display manager that runs on wayland like SDDM which i suspect is light enough to be started fully in 10 seconds.
I suggest changing the default values for `TimeoutStartSec` for the relevant services and Environmental variables to `30` or adding a hardware test for setting the values because many people aren't able to troubleshoot the problem. 
I will experiment with this for a few days to be sure it's solved and let people discuss it here if their problem is not solved . then i will close it with a comment to give them steps to solve this. 
Thanks everyone for great work. 
My relevant hardware info is : 
```md
**CPU:** Intel(R) Core(TM) i5-3337U CPU @ 1.80GHz
**STORAGE:** HDD 5400 RPM
```

**Comment by ArtemSmaznov** [comment]
I've added the overrides as suggested:
```cat .config/systemd/user/wayland-*/override.conf
TimeoutStartSec=60
TimeoutStartSec=60
TimeoutStartSec=60

ls .config/systemd/user/wayland-*/override.conf
.config/systemd/user/wayland-session-waitenv.service.d/override.conf      
.config/systemd/user/wayland-wm@.service.d/override.conf
.config/systemd/user/wayland-wm@hyprland.desktop.service.d/override.conf
```

and tried launching via `export UWSM_WAIT_VARNAMES_TIMEOUT=60 && uwsm start hyprland.desktop & top -b > ~/top.log` with no luck. Still getting the same crash. I'll try with a larger timeout and see if I can get this to work.

Also need to figure out how to analyze the top output. The only things that stand out to me is that I get 45.2% CPU for hyprland at one point and that I get 6 top outputs in the file for some reason.

I also have a `HDD 5400 RPM` and I also have `dm-crypt` encryption with `mount at login` - could my encryption setup cause additional delays? Or people normally have their home folder on an SSD? I do have quite a slow first login that I want to speed up if I can. Hyprland launches fast enough on second launch attempt but most applications like Qutebrowser and Emacs are unresponsive for a few minutes after that. I guess the system is loading stuff from my HDD to memory.

My system is quite beefy with `CPU: AMD Ryzen 9 7900X3D (24) @ 5.66 GHz` so I don't think that is a factor here.

**Comment by Vladimir-csp** [comment]
With 0.23.2 there should be no need for custom timeout overrides, they are placed automatically if `UWSM_WAIT_VARNAMES_TIMEOUT` is properly set. 

So 60 seconds is not enough?

**Comment by Random-Penguin** [comment]
Maybe the compositor is only responsive _after_ uwsm fails? I don't see another way this could happen... Then again, it's your project, not mine.

**Comment by ShadowS0ng** [comment]
> > `HDD 5400 RPM`
> 
> This venerable guy looks like the culprit. How long does the compositor's cold startup take on your machine?
> 
> You can set `StatusUnitFormat=combined` in `systemd/system.conf` to show unit IDs in Starting/Stopping messages, easier to keep track of them.

So i suppose it would be from the appearance of `uwsm_hyprland.desktop[pid]: Starting: Hyprland -c ...` to 
`systemd[pid]: Reached target Session of hyprland.desktop Wayland compositor` which it takes **17 seconds** to complete then it starts *daemons* and *exec-once* lines which takes 62 seconds. 
Below are `systemd-analyze --user blame` and `systemd-analyze critical-chain` Output:

```shell
# systemd-analyze --user blame: 
16.599s wayland-wm@hyprland.desktop.service
16.505s wayland-session-waitenv.service
11.019s random-wallpaper.service
 8.052s xdg-desktop-portal.service
 1.346s swww.service
  623ms app-nm\x2dapplet@autostart.service
  465ms xdg-document-portal.service
  427ms wayland-wm-env@hyprland.desktop.service
```
```shell 
# systemd-analyze --user critical-chain wayland-wm@hyprland.desktop.service :
wayland-wm@hyprland.desktop.service +16.599s
└─basic.target @1.420s
  └─systemd-tmpfiles-setup.service @1.374s +45ms
    └─app.slice @1.368s
      └─-.slice @1.264s
```
```shell
# systemd-analyze --user critical-chain graphical-session.target
graphical-session.target @27.386s
└─basic.target @1.420s
  └─systemd-tmpfiles-setup.service @1.374s +45ms
    └─app.slice @1.368s
      └─-.slice @1.264s
```

Let me know if you want it to be tested without UWSM and with default config also.

**Comment by Vladimir-csp** [comment]
From `Starting wayland-wm@...` to `Started wayland-wm@...` (assuming `StatusUnitFormat` is set to show IDs).

First message from uwsm inside unit and target reaching message should be roughly near those. So 17 seconds it is.

Does everything work if timeout is set larger than that?

**Comment by ShadowS0ng** [comment]
> From `Starting wayland-wm@...` to `Started wayland-wm@...` (assuming `StatusUnitFormat` is set to show IDs).
> 
> First message from uwsm inside unit and target reaching message should be roughly near those. So 17 seconds it is.
> 
> Does everything work if timeout is set larger than that?

Yes that's basically what happens, when i set the `TimeoutStartSec=` to `18` it works perfectly and then when i set it to `16` it fails.

**Comment by ArtemSmaznov** [comment]
> If still on 0.23.0, then fix your `override.conf` files to have `[Service]` section.
> 
> ```
> [Service]                                                                                    
> TimeoutStartSec=30
> ```

OMG I feel so silly, of course I needed to add the section. 

30 sec works for me flawlessly for me and hyprland starts in 11 sec or so. Can also confirm that setting `UWSM_WAIT_VARNAMES_TIMEOUT` is enought on `0.23.2`.

Is there any downsides to a longer timeout? 30 sec honestly seems like a pretty reasonable default as I don't imagine it holds anything.

**Comment by Random-Penguin** [comment]
Latest git works perfectly, thanks!
